File: go/internal/aider/edit_block.go
package aider

// Re-implementation of Aider's SEARCH/REPLACE diff format patch application

import (
	"regexp"
	"strings"
	"fmt"
)

type EditBlock struct {
	Search  string
	Replace string
}

func ParseEditBlocks(llmResponse string) []EditBlock {
	var blocks []EditBlock
	re := regexp.MustCompile(`(?s)<<<<<<< SEARCH\n(.*?)=======\n(.*?)>>>>>>> REPLACE`)
	matches := re.FindAllStringSubmatch(llmResponse, -1)

	for _, match := range matches {
		if len(match) == 3 {
			blocks = append(blocks, EditBlock{
				Search:  match[1],
				Replace: match[2],
			})
		}
	}
	return blocks
}

func ApplyEditBlocks(fileContent string, blocks []EditBlock) string {
	newContent := fileContent
	for _, block := range blocks {
		if strings.Contains(newContent, block.Search) {
			newContent = strings.Replace(newContent, block.Search, block.Replace, 1)
		} else {
			fmt.Println("Warning: Could not find SEARCH block exactly.")
		}
	}
	return newContent
}

File: csharp/AiderFeatures/EditBlock.cs
// Re-implementation of Aider's SEARCH/REPLACE diff format patch application

using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace TormentNexus.AiderFeatures
{
    public class EditBlock
    {
        public string Search { get; set; }
        public string Replace { get; set; }
    }

    public static class EditBlockParser
    {
        public static List<EditBlock> ParseEditBlocks(string llmResponse)
        {
            var blocks = new List<EditBlock>();
            var regex = new Regex(@"<<<<<<< SEARCH\n([\s\S]*?)=======\n([\s\S]*?)>>>>>>> REPLACE");
            var matches = regex.Matches(llmResponse);

            foreach (Match match in matches)
            {
                blocks.Add(new EditBlock
                {
                    Search = match.Groups[1].Value,
                    Replace = match.Groups[2].Value
                });
            }
            return blocks;
        }

        public static string ApplyEditBlocks(string fileContent, List<EditBlock> blocks)
        {
            string newContent = fileContent;
            foreach (var block in blocks)
            {
                if (newContent.Contains(block.Search))
                {
                    newContent = newContent.Replace(block.Search, block.Replace);
                }
                else
                {
                    Console.WriteLine("Warning: Could not find SEARCH block exactly.");
                }
            }
            return newContent;
        }
    }
}
